File: TST.py
############################################################
# TST.py
# Implementaion of ternary-search-trie. 
############################################################

import sys 
import random


def main():
    testTST()


class Node:

    def __init__(self, c=None):
        self.c = c
        self.val = None
        self.left = None 
        self.mid = None
        self.right = None

class TST:

    def __init__(self):
        self.root = Node()

    # ------------------------------------------------------
    # Inserting into trie
    # ------------------------------------------------------
    def put(self, key, val):
        """
        Inserts key into trie and associates key with value. 
        """
        assert(isinstance(key, str))
        if len(key) >= 1:
            index = 0
            self.root.mid = self.put_recursive(self.root.mid, key, val, index)

    def put_recursive(self, node, key, val, index):
        if node is None:
            node = Node(key[index])
        if key[index] < node.c:
            node.left = self.put_recursive(node.left, key, val, index) 
        elif key[index] > node.c:
            node.right = self.put_recursive(node.right, key, val, index) 
        elif index < len(key) - 1:
            node.mid = self.put_recursive(node.mid, key, val, index+1)
        else:
            node.val = val
        return node

    # ------------------------------------------------------
    # Retrieving from trie
    # ------------------------------------------------------
    def get(self, key):
        """
        Returns value associated with key, returns None if key 
        is not associated with any value in the trie. 
        """
        if self.root.mid is None:  # trie is empty 
            return None
        index = 0
        return self.get_recursive(self.root.mid, key, index)
    
    def get_recursive(self, node, key, index):
        if node is None:  # key is not in tree
            return None 
        if key[index] < node.c:
            return self.get_recursive(node.left, key, index)
        elif key[index] > node.c:
            return self.get_recursive(node.right, key, index)
        elif index < len(key) - 1:
            return self.get_recursive(node.mid, key, index+1)
        else:
            return node.val

    # ------------------------------------------------------
    # Deleting from trie
    # ------------------------------------------------------
    def delete(self, key): 
        """
        Deletes the given key from a trie. 
        """
        index = 0
        self.root.mid = self.delete_recursive(self.root.mid, key, index)
    
    def delete_recursive(self, node, key, index):
        if node is None:  # key is not in tree
            return None 
        if key[index] < node.c:
            node.left = self.delete_recursive(node.left, key, index) 
        elif key[index] > node.c:
            node.right = self.delete_recursive(node.right, key, index) 
        elif index < len(key) - 1:
            node.mid = self.delete_recursive(node.mid, key, index+1)
        else:
            node.val = None

        # if there are no more nodes down the middle path, check if we can delete this node. 
        # we can delete the middle node if there are no left or right paths as well as the 
        # middle path. 
        if node.val is None:
            if node.mid is None:
                if node.left is None: 
                    return node.right  # could be None, should be correct either way
                elif node.right is None: # we know that node.left is not None
                    return node.left

        return node


# ----------------------------------------------------------
# test functions
# ----------------------------------------------------------

def testTST(ntrials=1000, Llim=100):
    """
    Performs ntrials random operations in parallel in python dictionary and
    TST, in order to ensure that TST is implemented properly.

    Keys are no longer than Llim. 
    """
    A = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()-_=+<>?:\"';"

    arrkeys = list()

    delkeys = set()

    test_d = {}
    tst = TST()

    # a key is deleted with probability 10%
    for j in range(ntrials):
        u = random.uniform(0,1)
        L = random.randint(1, Llim)
        n = random.randint(-Llim, Llim)
        if u < 0.50:
            if len(arrkeys) > 0:
                i = random.randint(0, len(arrkeys)-1)
                s = arrkeys[i] 
                arrkeys.pop(i)
            else:
                s = random_string(L, A)
            if s in test_d:
                del test_d[s]
            tst.delete(s)

            delkeys.add(s)

            assert(s not in test_d and tst.get(s) is None)

        s = random_string(L, A)
        tst.put(s, n)
        test_d[s] = n

    del s
    for s in test_d:
        print(s, test_d[s], tst.get(s), s in delkeys)
        assert(test_d[s] == tst.get(s))

    print("all tests passed!", len(delkeys))
            


def random_string(L, alphabet):
    """
    Creates a random string of length L from a alphabet. 
    """
    k = len(alphabet)
    s = [None] * L 
    for i in range(L):
        u = random.uniform(0, 1)
        j = random.randint(0, k-1)
        if u < 0.5:
            s[i] = alphabet[j]
        else:
            s[i] = alphabet[j].lower()
    return ''.join(s)


if __name__ == '__main__':
    main()

File: TST.c
/* Implementation of a Ternary Search Trie with support for
insert, delete, and retrieval. Assumes non-negative value -
outputs -1 for queries not stored in data structure. */

#include <stdio.h>
#include <stdlib.h>

// node class for TST
typedef struct node
{
    char *c;
    int *val;
    struct node *left;
    struct node *mid;
    struct node *right;
} node;

// TST
typedef struct TST
{
    node *root;
} TST;

// functions for operating on TST
TST *newTree(void);
TST *put(TST *t, char *key, int val);
node *putRecursive(node *n, char *key, int val, int index);
int get(TST *t, char *key);
int getRecursive(node *n, char *key, int index);
TST *del(TST *t, char *key);
node *delRecursive(node *n, char *key, int index);
TST *delTree(TST *t);
void delNode(node *n);


int main(void)
{
    TST *t;
    t = newTree();
    if (t == NULL) {
        printf("Error: tree not successfully created in main!\n");
        return 1;
    }

    int R = 63;
    char *alph = "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890'abcdefghijklmnopqrstuvwxyz";

    // create nTrials new strings, associate with random non-negative value < Lim
    int nTrials = 10;
    int lim = 100;

    int trueVal;
    char *s[nTrials];
    for (int i = 0; i < nTrials; i++) {
        trueVal = rand() % lim;
        int len = rand() % 20;
        printf("The true value is: %i\n", trueVal);

        // create random string
        s[i] = malloc((len+1) * sizeof(char));
        if (s[i] == NULL) {
            printf("Error creating string!\n");
            return 1;
        }
        for (int j = 0; j < len; j++) {
            int u = rand() % R;
            s[i][j] = alph[u];
        }
        s[i][len] = '\0';

        // insert into TST
        put(t, s[i], trueVal);

        // get from TST
        int gotVal = get(t, s[i]);

        // print result
        printf("String: %s, Got value: %i, True value: %i\n", s[i], gotVal, trueVal);
        printf("----------------------------------------------\n");
    }

    for (int i = 1; i < nTrials; i++) {
        t = del(t, s[i]);
    }

    for (int i = 0; i < nTrials; i++) {
        int gotVal = get(t, s[i]);
        printf("String: %s, Got value: %i\n", s[i], gotVal);
    }

    char *missingFirst = "ello world";
    char *hello = "hello world";

    put(t, missingFirst, 123);
    int gotVal = get(t, &hello[1]);

    printf("String: %s, got value: %i (should be 123)\n", missingFirst, gotVal);

    char *empty = "";
    put(t, empty, 456);
    gotVal = get(t, empty);

    printf("Empty string, got value: %i (should be 456)\n", gotVal);
    t = del(t, "");
    gotVal = get(t, "");
    printf("Empty string after deleting, got value: %i (should be -1)\n", gotVal);
    for (int i = 0; i < nTrials; i++) {
        free(s[i]);
    }

    put(t, "HAZELLE", 123);
    printf("Value of HAZELLE (should be 123): %i\n", get(t, "HAZELLE"));

    t = delTree(t);
}


TST *newTree(void)
{
    TST *t = malloc(sizeof(TST));
    if (t == NULL) {
        printf("Error in function newTree:\n");
        printf("Error: failed to create tree!\n");
        exit(1);
    }
    t->root = NULL;
    return t;
}


TST *put(TST *t, char *key, int val)
{
    int index = 0;
    if (t->root == NULL) {
        t->root = malloc(sizeof(node));
        if (t->root == NULL) {
            printf("Error in function put:\n");
            printf("Error: failed to create node!\n");
            exit(1);
        }
        t->root->c = NULL;
        t->root->val = NULL;
        t->root->left = NULL;
        t->root->mid = NULL;
        t->root->right = NULL;
    }
    if (key[0] == '\0') {
        if (t->root->val == NULL) {
            t->root->val = malloc(sizeof(int));
        }
        *(t->root->val) = val;
    } else {
        t->root->mid = putRecursive(t->root->mid, key, val, index);
    }
    return t;
}


node *putRecursive(node *n, char *key, int val, int index)
{
    if (n == NULL) {
        n = malloc(sizeof(node));
        if (n == NULL) {
            printf("Error in function putRecursive: \n");
            printf("Error: failed to create node!\n");
            exit(1);
        }
        n->c = malloc(sizeof(char));
        if (n->c == NULL) {
            printf("Error in function putRecursive: \n");
            printf("Error: failed to allocate memory!\n");
            exit(1);
        }
        n->val = NULL;
        n->mid = NULL;
        n->left = NULL;
        n->right = NULL;
        *(n->c) = key[index];
    }
    if (key[index] < *(n->c)) {
        node *newNode;
        newNode = putRecursive(n->left, key, val, index);
        n->left = newNode;
    } else if (key[index] > *(n->c)) {
        node *newNode;
        newNode = putRecursive(n->right, key, val, index);
        n->right = newNode;
    } else if (key[index+1] != '\0') {
        node *newNode;
        newNode = putRecursive(n->mid, key, val, index+1);
        n->mid = newNode;
    } else {
        if (n->val == NULL) {
            n->val = malloc(sizeof(int));
            if (n->val == NULL) {
                printf("Error in function putRecursive:\n");
                printf("Error: failed to allocate memory for integer!\n");
                exit(1);
            }
        }
        *(n->val) = val;
    }
    return n;
}


int get(TST *t, char *key)
{
    if (t->root == NULL) {
        return -1;
    }
    // handle case of empty string
    if (key[0] == '\0') {
        if (t->root->val == NULL) {
            return -1;
        } else {
            return *(t->root->val);
        }
    }
    int index = 0;
    return getRecursive(t->root->mid, key, index);
}

int getRecursive(node *n, char *key, int index)
{
    if (n == NULL) {
        return -1;
    }
    if (key[index] < *(n->c)) {
        return getRecursive(n->left, key, index);
    } else if (key[index] > *(n->c)) {
        return getRecursive(n->right, key, index);
    } else if (key[index+1] != '\0') {
        return getRecursive(n->mid, key, index+1);
    } else {
        int out;
        // note that we need to check for short prefixes
        if (n->val == NULL) {
            out = -1;
        } else {
            out = *(n->val);
        }
        return out;
    }
}

/* Delete key from tree */
TST *del(TST *t, char *key)
{
    // handle passage of empty string
    if (key[0] == '\0') {
        if (t->root != NULL && t->root->val != NULL) {
            free(t->root->val);
            t->root->val = NULL;
        }
        return t;
    }
    int index = 0;
    if (t->root->mid == NULL) {
        return t;
    }
    t->root->mid = delRecursive(t->root->mid, key, index);
    return t;
}

node *delRecursive(node *n, char *key, int index)
{
    if (n == NULL) {
        return NULL;
    }
    if (key[index] < *(n->c)) {
        n->left = delRecursive(n->left, key, index);
    } else if (key[index] > *(n->c)) {
        n->right = delRecursive(n->right, key, index);
    } else if (key[index+1] != '\0') {
        n->mid = delRecursive(n->mid, key, index+1);
    } else {
        free(n->val);
        n->val = NULL;
    }


    /* if there are no more nodes down the middle path, check if we can delete this node.
       we can delete the middle node if there are no left or right paths as well as the
       middle path. */

    if (n->val == NULL && n->mid == NULL) {
        if (n->left == NULL && n->right == NULL) {
            free(n->c);
            free(n);
            return NULL;
        } else if (n->left == NULL) {
            node *rightNode = n->right;
            free(n->c);
            free(n);
            return rightNode;
        } else if (n->right == NULL) {
            node *leftNode = n->left;
            free(n->c);
            free(n);
            return leftNode;
        }
    }

    return n;
}

/* Deletes entire tree */
TST *delTree(TST *t)
{
    node *root = t->root;
    delNode(root);
    root = NULL;
    free(t);
    t = NULL;
    return t;
}

void delNode(node *n)
{
    if (n == NULL) {
        return;
    }

    if (n->left != NULL) delNode(n->left);
    if (n->mid != NULL) delNode(n->mid);
    if (n->right != NULL) delNode(n->right);

    free(n->c);
    free(n->val);
    free(n);
}
